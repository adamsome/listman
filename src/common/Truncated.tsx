/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import React from 'react'

export type TruncatedProps = typeof defaultProps & {
  children: React.ReactNode
}

const defaultProps = {
  lines: 1 as number,
}

export const Truncated = (props: TruncatedProps) => {
  const { lines } = props
  if (lines < 1) {
    return <div {...props} />
  }
  if (lines === 1) {
    return <TruncatedLine {...props} />
  }
  return <ClampedLines {...props} />
}

const TruncatedLine = (props: TruncatedProps) => {
  return (
    <div
      css={css`
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
      `}
    >
      {props.children}
    </div>
  )
}

const ClampedLines = (props: TruncatedProps) => {
  const { lines, children } = props
  return (
    <div
      css={css`
        display: -webkit-box;
        -webkit-line-clamp: ${lines};
        -webkit-box-orient: vertical;
        overflow: hidden;
      `}
    >
      {children}
    </div>
  )
}

Truncated.defaultProps = defaultProps

export default Truncated
