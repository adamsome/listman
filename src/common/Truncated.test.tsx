import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Truncated from './Truncated'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Truncated lines={2}>text</Truncated>, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const tree = renderer.create(<Truncated lines={2}>text</Truncated>).toJSON()
  expect(tree).toMatchSnapshot()
})
