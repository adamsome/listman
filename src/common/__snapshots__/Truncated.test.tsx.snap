// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

<div
  className="emotion-0"
>
  text
</div>
`;
